listings/applications/campaign_factory/src/contract.cairo

```diff
@@ -0,0 +1,153 @@
+        pub contract_address: ContractAddress
+    }
+
+    pub mod Errors {
+        pub const CLASS_HASH_ZERO: felt252 = 'Class hash cannot be zero';
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, class_hash: ClassHash) {
+        assert(class_hash.is_non_zero(), Errors::CLASS_HASH_ZERO);
+        self.campaign_class_hash.write(class_hash);
+        self.ownable._init(get_caller_address());
+    }
+
+
+    #[abi(embed_v0)]
+    impl CampaignFactory of super::ICampaignFactory<ContractState> {
+        // ANCHOR: deploy
+        fn create_campaign(
+            ref self: ContractState,
+            title: ByteArray,
+            description: ByteArray,
+            target: u256,
+            duration: u64,
+            token_address: ContractAddress,
+        ) -> ContractAddress {
+            let caller = get_caller_address();
+
+            // Create contructor arguments
+            let mut constructor_calldata: Array::<felt252> = array![];
+            ((caller, title, description, target), duration, token_address)
+                .serialize(ref constructor_calldata);
+
+            // Contract deployment
+            let (contract_address, _) = deploy_syscall(
+                self.campaign_class_hash.read(), 0, constructor_calldata.span(), false
+            )
+                .unwrap_syscall();
+
+            // track new campaign instance
+            let mut campaigns = self.campaigns.read();
+            campaigns.append(ICampaignDispatcher { contract_address }).unwrap();
+
+            self.emit(Event::CampaignCreated(CampaignCreated { caller, contract_address }));
+
+            contract_address
+        }
+        // ANCHOR_END: deploy
+
+        fn get_campaign_class_hash(self: @ContractState) -> ClassHash {
+            self.campaign_class_hash.read()
+        }
+
+        fn update_campaign_class_hash(ref self: ContractState, new_class_hash: ClassHash) {
+            self.ownable._assert_only_owner();
+
+            // update own campaign class hash value
+            self.campaign_class_hash.write(new_class_hash);
+
+            // upgrade each campaign with the new class hash
+            let campaigns = self.campaigns.read();
+            let mut i = 0;
+            while let Option::Some(campaign) = campaigns
+                .get(i)
+                .unwrap_syscall() {
+                    if let Result::Err(errors) = campaign.upgrade(new_class_hash) {
+                        self
+                            .emit(
+                                Event::ClassHashUpdateFailed(
+                                    ClassHashUpdateFailed {
+                                        campaign: campaign.contract_address, errors
+                                    }
+                                )
+                            )
+                    }
+                    i += 1;
+                };
+
+            self.emit(Event::ClassHashUpdated(ClassHashUpdated { new_class_hash }));
```

Comment:
Do you want to continue and emit `ClassHashUpdated` even if there's one of the contract upgrade that failed?
The state of the contract might be incoherent, is the following an invariant you want to enforce?
$\forall c \in self.campaigns ~ | ~ c.classHash == self.campaignClassHash$

Reply:
I'd like that, but then we halt all upgrades when just one of them fails.

Anyway, we might be able to solve this by moving the upgrade responsibility onto campaign creators, see https://github.com/NethermindEth/StarknetByExample/pull/219#discussion_r1636281869